clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java

```diff
@@ -2306,77 +2325,103 @@ void handleResponse(AbstractResponse abstractResponse) {
+                return new MetadataRequest.Builder(Collections.singletonList(Topic.GROUP_METADATA_TOPIC_NAME), true);
             }
 
             @Override
             void handleResponse(AbstractResponse abstractResponse) {
                 MetadataResponse metadataResponse = (MetadataResponse) abstractResponse;
 
-                final Map<Node, KafkaFutureImpl<Collection<ConsumerGroupListing>>> futures = new HashMap<>();
-
-                for (final Node node : metadataResponse.brokers()) {
-                    futures.put(node, new KafkaFutureImpl<Collection<ConsumerGroupListing>>());
+                for (final MetadataResponse.TopicMetadata metadata : metadataResponse.topicMetadata()) {
+                    if (metadata.topic().equals(Topic.GROUP_METADATA_TOPIC_NAME)) {
+                        for (final MetadataResponse.PartitionMetadata partitionMetadata : metadata.partitionMetadata()) {
+                            final Node leader = partitionMetadata.leader();
+                            if (partitionMetadata.error() != Errors.NONE) {
+                                // TODO: KAFKA-6789, retry based on the error code
+                                KafkaFutureImpl<Collection<ConsumerGroupListing>> future = new KafkaFutureImpl<>();
+                                future.completeExceptionally(partitionMetadata.error().exception());
+                                // if it is the leader not found error, then the leader might be NoNode; if there are more than
+                                // one such error, we will only have one entry in the map. For now it is okay since we are not
+                                // guaranteeing to return the full list of consumers still.
+                                futuresMap.put(leader, future);
+                            } else {
+                                futuresMap.put(leader, new KafkaFutureImpl<Collection<ConsumerGroupListing>>());
+                            }
+                        }
+                        listFuture.complete(null);
+                    } else {
+                        if (metadata.error() != Errors.NONE)
+                            listFuture.completeExceptionally(metadata.error().exception());
+                        else
+                            listFuture.completeExceptionally(new IllegalStateException("Unexpected topic metadata for "
+                                    + metadata.topic() + " is returned; cannot find the brokers to query consumer listings."));
+                    }
                 }
 
-                future.combine(futures.values().toArray(new KafkaFuture[0])).thenApply(
-                        new KafkaFuture.BaseFunction<Collection<ConsumerGroupListing>, Collection<ConsumerGroupListing>>() {
+                // we have to flatten the future here instead in the result, because we need to wait until the map of nodes
+                // are known from the listNode request.
+                flattenFuture.copyWith(
```

Comment:
Unfortunately we cannot, because we need to construct a final future variable at the beginning that can be returned at the end of the call, and this future can only be "initialized" with the underlying map after the first round trip. `thenApply` will return a new future, which cannot be used here.